packages/components/src/data-model/word-query/word-query.js

```diff
@@ -0,0 +1,314 @@
+
+/**
+ * Contains all the business logic that is necessary to resolve the GraphQL word query.
+ */
+export default class WordQuery {
+  constructor (options = {}, queryParams = {}) {
+    this._context = {
+      queryParams,
+      options: {
+        language: Language.fromJsonObject({ code: queryParams.language }),
+        clientID: options.clientID,
+        lexicons: options.lexicons,
+        shortLexicons: options.shortLexicons
+      },
+      cancelled: false,
+      result: undefined
+    }
+  }
+
+  start (result) {
+    this._context.result = result
+    this._context.result.loadingStarted()
+    if (this._context.queryParams.getLexemes) { this._context.result.state.lexemes.loading = true }
+    if (this._context.queryParams.getShortDefs) { this._context.result.state.shortDefs.loading = true }
+    const executionPath = WordQuery._createExecutionPath(this._context.queryParams)
+    try {
+      // Ignore result of an asynchronous operation and return immediately
+      // Result will be tracked via the proxied `result` variable
+      this._context = WordQuery._execute(executionPath, this._context)
+    } catch (err) {
+      Logger.getInstance().error(err)
+    }
+    return result
+  }
+
+  cancel () {
+    this._context.cancelled = true
+  }
+
+  // Creates a set of methods to be executed in order to fulfil the request.
+  static _createExecutionPath (queryParams) {
+    /*
+     Execution path consists of several batches. Batches are execution sequentially: each next batch
+     is executed after the previous one is completed.
+     Each batch consist of one or several jobs. Jobs within a batch can be executed in parallel.
+     */
+    let path = [] // eslint-disable-line prefer-const
+
+    let getLexemesBatch = [] // eslint-disable-line prefer-const
+    if (queryParams.getLexemes) {
+      if (queryParams.useMorphService) {
+        getLexemesBatch.push('_getTuftsMorphology')
+      }
+      if (queryParams.useTreebankData) {
+        getLexemesBatch.push('_getTreebankData')
+      }
+      if (queryParams.useWordAsLexeme) {
+        getLexemesBatch.push('_getWordAsLexeme')
+      }
+    }
+    if (getLexemesBatch.length > 0) {
+      path.push(getLexemesBatch)
+    }
+    path.push(['_disambiguate'])
+
+    if (queryParams.getShortDefs) {
+      path.push(['_getShortDefs'])
+    }
+    return path
+  }
+
+  // Runs all the jobs within the execution path.
+  static async _execute (executionPath, context) {
+    try {
+      for (const batch of executionPath) {
+        if (!context.cancelled) {
+          const jobs = batch.map(job => WordQuery[job](context))
+          await Promise.all(jobs)
+        }
+      }
+    } catch (err) {
+      Logger.getInstance().error(err)
+    }
+    WordQueryResult.loadingStopped(context.result.state)
+    context.result.state = Object.assign({}, context.result.state)
+    return context
+  }
+
+  // Retrieves morphological data from Tufts
+  static async _getTuftsMorphology (context) {
```

Comment:
do these methods `_getTuftsMorphology`, `_disambiguate` and others should be considered as query steps? graphQl results in the execution path? 
If it is so I suggest to move it to a separate file , because it should be updated independently from the query construction process. what do you think?

Reply:
Yes, those methods are the query steps. I think splitting them into a separate file is a very good idea because it would increase a separation of concerns. @balmas, would you agree with that?

Reply:
yes